iModelCore/ECPresentation/Source/Shared/RulesPreprocessor.cpp

```diff
@@ -978,14 +978,53 @@ bvector<ContentRuleCP> RulesPreprocessor::_GetContentRules()
     return result;
     }
 
+/*---------------------------------------------------------------------------------**//**
+* @bsimethod
++---------------+---------------+---------------+---------------+---------------+------*/
+static bvector<NavNodeKeyCPtr> SplitLabelGroupedKeys(IConnectionCR connection, INodeInstanceKeysProvider const* instanceKeyProvider, LabelGroupingNodeKey const* key)
+    {
+    bvector<NavNodeKeyCPtr> result;
+    bmap<ECN::ECClassId, bvector<ECInstanceKey>> instanceKeyMap;
+    instanceKeyProvider->IterateInstanceKeys(*key,
+        [&instanceKeyMap](ECInstanceKey instanceKey)
+            {
+                if (instanceKeyMap[instanceKey.GetClassId()].empty())
+                    instanceKeyMap[instanceKey.GetClassId()] = { instanceKey };
+                else
+                    instanceKeyMap[instanceKey.GetClassId()].push_back(instanceKey);
+                return true;
+            });
+    for (auto& instanceKeys : instanceKeyMap)
+        {
+        std::unique_ptr<bvector<ECInstanceKey>> keyList = std::make_unique<bvector<ECInstanceKey>>(instanceKeys.second);
+        result.push_back(LabelGroupingNodeKey::Create(connection, key->GetSpecificationIdentifier(),
+            nullptr, key->GetLabel(), instanceKeys.second.size(), nullptr, std::move(keyList)));
+        }
+    return result;
+    }
+
 /*---------------------------------------------------------------------------------**//**
 * @bsimethod
 +---------------+---------------+---------------+---------------+---------------+------*/
 ContentRuleInputKeysContainer RulesPreprocessor::_GetContentSpecifications(ContentRuleParametersCR params)
     {
     ContentRuleInputKeysContainer specs;
     bset<NavNodeKeyCP> handledNodes;
+    bvector<NavNodeKeyCPtr> singleNodeKeys;
     for (NavNodeKeyCPtr const& inputNodeKey : params.GetInputNodeKeys())
+        {
+        if (inputNodeKey->GetType().Equals("DisplayLabelGroupingNode"))
+            {
+            bvector<NavNodeKeyCPtr> keys = SplitLabelGroupedKeys(m_connection, params.GetInstanceKeyProvider(), inputNodeKey->AsLabelGroupingNodeKey());
+            singleNodeKeys.insert(singleNodeKeys.end(), keys.begin(), keys.end());
+            }
+        else
+            {
+            singleNodeKeys.push_back(inputNodeKey);
+            }
+        }

```

Comment:
I would suggest extracting this code into separate function.

Reply:
Also I guess we do not need to try splitting keys if `GetInstanceKeyProvider` returns `nullptr`